### PR TITLE
Add feature to use NUSMods data

### LINKED ISSUE #59: Add NUSMods data functionality
As a user, I want to be able to access module data that is obtained from NUSMods within GradPad, so that I don't have to switch to my browser to access NUSMods.

### PR DESCRIPTION
**Pull request by Silvernitro**
This PR adds the functionality to read module data from NUSMods. 
It gets fresh data from the API if possible. Otherwise, it will still read the stale data stored locally (`modules.json`).

There are a few main components here:
- The `DataFetcher` interface represents operations to fetch data from the NUSMods public API. It is implemented by the `DataFetcherManager` class. Most of the time, you shouldn't need to interact with this class directly, because:

- The `NusmodsData` abstract class holds the public methods that the rest of GradPad can use to access module data. It is implemented by the `NusmodsDataManager` class.

- The `ModuleSummary` class represents module summaries received from the `/moduleList` NUSMods endpoint. This is not used by GradPad and is only an intermediate container for `DataFetcher` to use.

- The `ModuleInfo` class is the one that represents the useful info from NUSMods. See the class fields to know what info is available.

- In general, accessing module info should be done through the `NusmodsData` public interface. Given a module code, it can give you the module title, or `ModuleInfo` object. Example usage:
```
NusmodsData data = new NusmodsDataManager();
data.getModuleTitle("CS2103T") // --> returns "Software Engineering"
```

- The `HttpUtil` class handles HTTP GET requests. It uses the wrapper singleton pattern instead of static methods to facilitate dependency injection while testing.

**Notes**

- Test coverage for all the code here is 92% with all tests passing. Not 100% tho hehe 😅

### CODE REVIEW

File: src/test/data/NusmodsDataManagerTest/validModule.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "moduleCode" : "CS2103T",
+  "title" : "Software Engineering",
+  "moduleCredit" : "4",
+  "description" : "The best module there is.",
```
Comment (shaokiat) [inline src/test/data/NusmodsDataManagerTest/validModule.json:5]: LOL

File: src/main/java/seedu/address/nusmods/DataFetcherManager.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    /**
+     * Fetches a list of all NUS module summaries, then fetches module info for all CS modules (only) and
+     * saves them to a JSON file locally.
+     *
+     * @throws NusmodsException if an error occurs while fetching data from NUSMods API or while saving data.
+     */
+    public void fetchAndSaveModules() throws NusmodsException {
+        List<ModuleSummary> moduleSummaries = fetchModuleSummaryList();
+        List<ModuleSummary> csModules = filterModuleSummaries(moduleSummaries, "CS");
+        Map<String, ModuleInfo> moduleInfoMap = generateModuleInfoMap(csModules);
```
Comment (shaokiat) [inline src/main/java/seedu/address/nusmods/DataFetcherManager.java]: I'm guessing we have to add the other non-CS modules that are in the syllabus as well to the moduleInfoMap

**Review by shaokiat — APPROVED** [review]
💯 Really outdone all of us this time! MVP 🥇 
Learnt a lot from your code! Thanks Sam 👍

File: src/main/java/seedu/address/nusmods/DataFetcherManager.java (hunk 2 above)
Comment (Silvernitro) [reply src/main/java/seedu/address/nusmods/DataFetcherManager.java]: Oh yeah you're right. In that case, I'll let `filterModuleSummaries` take in a bunch of keywords to filter against. We can then define the keywords in a hardcoded constant depending on the list of required modules.

**Comment by Silvernitro** [comment]
Update: As per @shaokiat suggestion, tweaked `filterModuleSummaries` method in `DataFetcherManager` class to accept multiple keywords to filter against. For now, I put some sample keywords in a constant `MODULE_FILTER_KEYWORDS` in the same class. We can change it when we work on required modules?

**Review by yan-soon — APPROVED** [review]
Daddy sam back at it again with the CLEANEST code 😤😩😩🥵🥵💯💯💦💦 (Everything looks great!)

**Pull request merged**